### PR TITLE
OHRM5X-1358: develop candidates view & add screen

### PR DESCRIPTION
**Pull request by Ajiharan007**

### CODE REVIEW

File: src/plugins/orangehrmRecruitmentPlugin/Controller/ViewCandidateController.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    /**
+     * @inheritDoc
+     */
+    public function preRender(Request $request): void
+    {
+        $component = new Component('view-candidate');
+        $this->setComponent($component);
+    }
+
+}
```
Comment (Super-Chama) [inline src/plugins/orangehrmRecruitmentPlugin/Controller/ViewCandidateController.php:37]: Line end

File: src/plugins/orangehrmRecruitmentPlugin/config/routes.yaml [hunk 2]
```diff
@@ -1,4 +1,14 @@
...
+save_recruitment_candidates:
+  path: /recruitment/addCandidate
+  controller: OrangeHRM\Recruitment\Controller\SaveCandidateController::handle
+  methods: [ GET ]
+
 recruitment_module_under_development: # remove this
   path: /recruitment/{screen}
   controller: OrangeHRM\Core\Controller\Common\ModuleUnderDevelopmentController::handle
-  methods: [ GET ]
+  methods: [ GET ]
```
Comment (Super-Chama) [inline src/plugins/orangehrmRecruitmentPlugin/config/routes.yaml:14]: Line end

File: src/plugins/orangehrmRecruitmentPlugin/Controller/SaveCandidateController.php [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+{
+    /**
+     * @inheritDoc
+     */
+    public function preRender(Request $request): void
+    {
+        $component = new Component('save-candidate');
+        $this->setComponent($component);
+    }
+}
```
Comment (Super-Chama) [inline src/plugins/orangehrmRecruitmentPlugin/Controller/SaveCandidateController.php:36]: Line end

File: installer/Migration/V5_1_0/LangStringHelper.php [hunk 4]
```diff
@@ -0,0 +1,253 @@
+<?php
```
Comment (Super-Chama) [inline installer/Migration/V5_1_0/LangStringHelper.php:1]: @SharukaNP won't this be conflicting with your [PR](https://github.com/Super-Chama/orangehrm/pull/52)? Please advise how to resolve this ?

File: src/client/src/orangehrmRecruitmentPlugin/components/CandidateAutocomplete.vue [hunk 5]
```diff
@@ -0,0 +1,84 @@
+<!--
```
Comment (Super-Chama) [inline src/client/src/orangehrmRecruitmentPlugin/components/CandidateAutocomplete.vue:1]: Any reason for adding specific candidate autocomplete? can't we re-use employee autocomplete ?

File: src/client/src/orangehrmRecruitmentPlugin/pages/SaveCandidate.vue [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    <div class="orangehrm-card-container">
+      <oxd-form :loading="isLoading" @submitValid="onSave">
+        <oxd-form-row>
+          <oxd-grid :cols="1" class="orangehrm-full-width-grid">
+            <oxd-grid-item>
+              <full-name-input
+                v-model:first-name="candidate.firstName"
+                v-model:middle-name="candidate.middleName"
+                v-model:last-name="candidate.lastName"
+                :rules="rules"
```
Comment (Super-Chama) [inline src/client/src/orangehrmRecruitmentPlugin/pages/SaveCandidate.vue:31]: required?

File: src/client/src/orangehrmRecruitmentPlugin/pages/SaveCandidate.vue [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        resume: '',
+        vacancy: null,
+        keywords: '',
+        application: '',
+        notes: '',
+        keep: '',
+      },
+      rules: {
+        firstName: [required, shouldNotExceedCharLength(30)],
+        lastName: [required, shouldNotExceedCharLength(30)],
```
Comment (Super-Chama) [inline src/client/src/orangehrmRecruitmentPlugin/pages/SaveCandidate.vue:170]: middle name length validation
Contact Number validation
resume validation
keywords validation
notes validation
date of application validation

File: src/client/src/orangehrmRecruitmentPlugin/pages/SaveCandidate.vue [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+        email: [required, validEmailFormat],
```
Comment (Super-Chama) [inline src/client/src/orangehrmRecruitmentPlugin/pages/SaveCandidate.vue:171]: email should have length validation

File: src/client/src/orangehrmRecruitmentPlugin/pages/SaveCandidate.vue [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+      },
+    };
+  },
+  methods: {
+    onSave() {
+      this.isLoading = true;
+      this.http.create({data: this.candidate}).then(() => {
+        this.isLoading = false;
```
Comment (Super-Chama) [inline src/client/src/orangehrmRecruitmentPlugin/pages/SaveCandidate.vue:179]: Please add success toast and redirection if applicable

**Review by Super-Chama — CHANGES_REQUESTED** [review]

File: installer/Migration/V5_1_0/LangStringHelper.php (hunk 4 above)
Comment (SharukaNP) [reply installer/Migration/V5_1_0/LangStringHelper.php:1]: Ah no I can rebase once his PR is merged vise-versa depending on whose PR is merged first

File: src/client/src/orangehrmRecruitmentPlugin/pages/ViewCandidate.vue [hunk 10]
```diff
@@ -0,0 +1,436 @@
...
+      {
+        id: 6,
+        label: $t('recruitment.job_offered'),
+      },
+      {
+        id: 7,
+        label: $t('recruitment.offered_declined'),
+      },
+    ];
+    const userdataNormalizer = data => {
```
Comment (Super-Chama) [inline src/client/src/orangehrmRecruitmentPlugin/pages/ViewCandidate.vue:222]: userdataNormalizer please rename

File: src/client/src/orangehrmRecruitmentPlugin/pages/ViewCandidate.vue [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+  name: 'ViewCandidate',
+  components: {
+    CandidateAutocomplete,
+    HiringManagerDropdown,
+    VacancyDropdown,
+    JobtitleDropdown,
+    'delete-confirmation': DeleteConfirmationDialog,
+  },
+  props: {
+    unselectableIds: {
```
Comment (Super-Chama) [inline src/client/src/orangehrmRecruitmentPlugin/pages/ViewCandidate.vue:185]: according to spec are there undeletable candidates?

File: installer/Migration/V5_1_0/Migration.php [hunk 12]
```diff
@@ -37,6 +42,41 @@ public function up(): void
...
         $this->getSchemaHelper()->addForeignKey('ohrm_kpi', $kpiForeignKeyConstraint);
+
+        $groups = ['recruitment'];
+        foreach ($groups as $group) {
+            $this->getLangStringHelper()->deleteNonCustomizedLangStrings($group);
+            $this->getLangStringHelper()->insertOrUpdateLangStrings($group);
+        }
+
+        $langCodes = [
+            'en_US',
```
Comment (SharukaNP) [inline installer/Migration/V5_1_0/Migration.php:53]: Aren't we adding the bg_BG translations? 
What about other languages?

File: src/client/src/orangehrmRecruitmentPlugin/pages/ViewCandidate.vue [hunk 13]
```diff
@@ -0,0 +1,436 @@
...
+  'v.vacancy': 'ASC',
+  'c.candidate': 'ASC',
+  'h.manager': 'ASC',
+  's.status': 'DEFAULT',
+};
+
+export default {
+  name: 'ViewCandidate',
+  components: {
+    CandidateAutocomplete,
```
Comment (Super-Chama) [inline src/client/src/orangehrmRecruitmentPlugin/pages/ViewCandidate.vue:178]: Lint issues? 
Please use kebab form like `'delete-confirmation': DeleteConfirmationDialog,`

File: src/client/src/orangehrmRecruitmentPlugin/pages/ViewCandidate.vue [hunk 14]
```diff
@@ -0,0 +1,436 @@
...
+            download: {
+              onClick: this.onClickEdit,
+              props: {
+                name: 'download',
+              },
+            },
+            view: {
+              onClick: this.onDownload,
+              props: {
+                name: 'menu-recruitment',
```
Comment (Super-Chama) [inline src/client/src/orangehrmRecruitmentPlugin/pages/ViewCandidate.vue:343]: please use correct icon

File: installer/Migration/V5_1_0/lang-string/recruitment.yaml [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    value: 'Date of Application'
+    unitId: date_of_application
+  -
+    value: 'Method of Application'
+    unitId: method_of_application
+  -
+    value: 'Enter comma se'
+    unitId: enter_comma_se
+  -
+    value: 'Candidate'
```
Comment (SharukaNP) [inline installer/Migration/V5_1_0/lang-string/recruitment.yaml:24]: There are translations for the word candidate in most languages. Please check again with the 4x xml files

File: src/client/src/orangehrmRecruitmentPlugin/pages/ViewCandidate.vue [hunk 16]
```diff
@@ -0,0 +1,436 @@
...
+        },
+      ],
+    };
+  },
+  methods: {
+    onClickAdd() {
+      navigate('/recruitment/addCandidate');
+    },
+    onClickEdit(item) {
+      navigate('/recruitment/addCandidate/{id}', {id: item.id});
```
Comment (Super-Chama) [inline src/client/src/orangehrmRecruitmentPlugin/pages/ViewCandidate.vue:366]: on click edit should go to edit path

**Review by Super-Chama — CHANGES_REQUESTED** [review]

**Review by SharukaNP — CHANGES_REQUESTED** [review]
Please check the translations again.

File: src/client/src/orangehrmRecruitmentPlugin/components/CandidateAutocomplete.vue (hunk 5 above)
Comment (Ajiharan007) [reply src/client/src/orangehrmRecruitmentPlugin/components/CandidateAutocomplete.vue:1]: yes we can

File: src/client/src/orangehrmRecruitmentPlugin/pages/SaveCandidate.vue (hunk 9 above)
Comment (Ajiharan007) [reply src/client/src/orangehrmRecruitmentPlugin/pages/SaveCandidate.vue:179]: not sure where to direct one

File: installer/Migration/V5_1_0/Migration.php (hunk 12 above)
Comment (Ajiharan007) [reply installer/Migration/V5_1_0/Migration.php:53]: yes we have

File: src/client/src/orangehrmRecruitmentPlugin/pages/ViewCandidate.vue (hunk 11 above)
Comment (Ajiharan007) [reply src/client/src/orangehrmRecruitmentPlugin/pages/ViewCandidate.vue:189]: i think no

File: src/client/src/core/util/validation/rules.ts [hunk 17]
```diff
@@ -73,6 +75,18 @@ export const validDateFormat = function(dateFormat = 'yyyy-MM-dd') {
   };
 };
 
+export const shouldBeCurrentOrPreviousDate = function() {
```
Comment (Super-Chama) [inline src/client/src/core/util/validation/rules.ts:78]: Need to add unit tests for this

Comment (Super-Chama) [reply src/client/src/core/util/validation/rules.ts:78]: Please check https://github.com/orangehrm/orangehrm/blob/main/src/client/src/core/util/validation/__tests__/rules.spec.ts

**Review by Super-Chama — APPROVED** [review]

**Pull request merged**